Infer: replace List reverse by ListBuffer

diesalbla opened the pull request:
In the `makeFullyDefined` method of the Infer file, instead of building a List by appending each element and then reversing the list, which builds (allocates) 2 lists, we now use a ListBuffer
and append in the right order, so we only allocate one list.

diesalbla committed:
Infer: replace List reverse by ListBuffer

In the `makeFullyDefined` method of the Infer file, instead of
building a List by appending each element and then reversing
the list, which builds (allocates) 2 lists, we now use a ListBuffer
and append in the right order, so we only allocate one list.

For convenience, and if the frequent case is to have no tparams at
all, we start `tparams` as `null`, and only allocate unless called at
least once.

retronym commented on src/compiler/scala/tools/nsc/typechecker/Infer.scala:
```diff
@@ -345,10 +346,10 @@ trait Infer extends Checkable {
      */
     def makeFullyDefined(tp: Type): Type = {
-      var tparams: List[Symbol] = Nil
+      val tparams: ListBuffer[Symbol] = ListBuffer.empty
```
I wonder what's the distribution of `tparams.length` at the end of this method? If `0` is the most common case, this change will be counter-productive.

You could leave this as `null` to start with, and allocate lazily on for first addition of a `tparam`.

diesalbla replied:
Done!

retronym reviewed (dismissed):
Measure how important the empty tparams case is, and add a zero-allocation path for that if needed.

hrhino commented on src/compiler/scala/tools/nsc/typechecker/Infer.scala:
```diff
@@ -345,10 +346,12 @@ trait Infer extends Checkable {
     def makeFullyDefined(tp: Type): Type = {
-      var tparams: List[Symbol] = Nil
+      var tparams_ : ListBuffer[Symbol] = null
+      def tparams: ListBuffer[Symbol] = { if (tparams_ == null) tparams_ = ListBuffer.empty ; tparams_ }
```
This should be (manually) inlined; as it stands, it'll make an `ObjectRef` wrapping `tparams_` by which to modify the var in an outer scope.

retronym replied:
I've pushed a commit to this PR to move `tparams_` into a field of the `TypeMap`.

hrhino commented on src/compiler/scala/tools/nsc/typechecker/Infer.scala:
```diff
@@ -345,10 +346,12 @@ trait Infer extends Checkable {
-      var tparams: List[Symbol] = Nil
+      var tparams_ : ListBuffer[Symbol] = null
+      def tparams: ListBuffer[Symbol] = { if (tparams_ == null) tparams_ = ListBuffer.empty ; tparams_ }
+      def tparamsList: List[Symbol] = if (tparams_ == null) Nil else tparams_.toList
```
ditto.

retronym committed:
Further refactor makeFullyDefined into a full TypeMap

We were already paying for the allocation of the capturing
lambda passed to `Type.map`, let's just make the `TypeMap`
apparent and use it to store the lazily contructed buffer
of existential type params.

While we're cleaning this up, move the name logic out
to `nme.existentialName`, and have that cache the names
for small values of `i`.
